In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/File/data_ggsheet.csv')

#MỤC LỤC:
#I. TỔNG QUAN:



##1.1. Dashboard Tổng quan: Row card các chỉ số tổng quan

Tổng doanh thu đạt 4.779 triệu VND với tổng số SKU là 92.346.Trong đó, 31.438 SKU đóng góp doanh thu chính, tức một phần SKU có thể không đóng góp hoặc đóng góp rất ít doanh thu.Điều này cho thấy có sự phân bổ không đồng đều giữa các SKU về mặt đóng góp doanh thu.

In [34]:
# Tính tổng Lượng bán
tong_luong_ban = df['SL'].sum()

# Tính tổng Doanh thu
tong_doanh_thu = df['Thành tiền'].sum()

# Tính tổng đơn hàng (đếm số đơn hàng duy nhất)
tong_don_hang = df['Mã đơn hàng'].nunique()

# Tạo DataFrame kết quả để hiển thị gọn
ket_qua = pd.DataFrame({
    'Tổng Lượng bán': [tong_luong_ban],
    'Tổng Doanh thu': [tong_doanh_thu],
    'Tổng Đơn hàng': [tong_don_hang]
})

ket_qua

Tổng Lượng bán  Tổng Doanh thu  Tổng Đơn hàng
0           92346      4779074000          31438

##1.2. Dashboard tổng quan: Doanh thu theo từng mặt hàng.

Tổng quan doanh thu theo từng sản phẩm ta thấy Sản phẩm [BOT01] Bột cần tây dẫn đầu với doanh thu 626 triệu VND và 15.650 SKUs, thể hiện rõ vai trò sản phẩm chủ lực. Theo sau là Trà nhuỵ hoa nghệ tây [TH006], Trà gừng [TTC01] và Trà hoa cúc trắng [TH003] với doanh thu từ 421 đến 592 triệu VND, cũng đóng góp lớn vào tổng doanh thu. Ngược lại, các sản phẩm dạng "Set", như Set trà hoa cúc trắng [SET03] hay Set trà gừng [SET04], có doanh thu thấp hơn rõ rệt, cho thấy hiệu quả kinh doanh chưa cao hoặc còn trong giai đoạn thử nghiệm. Biểu đồ này giúp dễ dàng xác định sản phẩm chiến lược và tiềm năng cần đầu tư thêm hay điều chỉnh.

In [35]:
# Nhóm theo 'Tên mặt hàng' và tính tổng doanh thu và lượng bán
df_summary = df.groupby('Tên mặt hàng').agg({
    'Thành tiền': 'sum',
    'SL': 'sum'
}).reset_index()

# Đổi tên cột cho dễ hiểu
df_summary = df_summary.rename(columns={
    'Thành tiền': 'Doanh thu',
    'SL': 'Lượng bán'
})

df_summary

Tên mặt hàng  Doanh thu  Lượng bán
0                        Bột cần tây  626000000      15650
1                            Cam lát  367380000       8164
2          Set 10 gói trà cam sả quế   50410000        710
3          Set 10 gói trà dưỡng nhan   98475000       1515
4        Set 10 gói trà gạo lứt 8 vị   31850000        910
5                Set 10 gói trà gừng   80000000       1600
6       Set 10 gói trà hoa cúc trắng  243900000       1626
7        Set 10 gói trà hoa đậu biếc  158712000       1167
8   Set 10 gói trà nụ hoa nhài trắng  114600000        955
9                     Trà cam sả quế  249888000       6576
10                    Trà dưỡng nhan  317170000       9062
11                  Trà gạo lứt 8 vị  130284000       7238
12                          Trà gừng  432900000      14430
13                     Trà hoa Atiso  144200000       2884
14                 Trà hoa cúc trắng  421440000       5268
15                  Trà hoa đậu biếc  279140000       4105
16             Trà nhuỵ hoa nghệ tây  592380000       3291
17          Trà nụ hoa hồng Tây Tạng  150315000       2733
18             Trà nụ hoa nhài trắng  290030000       4462

##1.3. Dashboard tổng quan: Doanh thu theo từng nhóm hàng.

Nhóm hàng lớn nhất về doanh thu là Trà hoa [THO] với 1.878 triệu VND, chiếm khoảng 39% tổng doanh thu.Tiếp theo là các nhóm:Trà củ, quả sấy [TTC] (800 triệu).Set trà [SET] (778 triệu).Trà mix [TMX] (697 triệu).Bột [BOT] (626 triệu).Nhóm Trà hoa chiếm ưu thế rõ ràng về doanh thu, cho thấy đây là nhóm hàng chủ lực, có thể là nhóm cần tập trung phát triển hoặc marketing nhiều hơn.
v

In [36]:
# Nhóm theo 'Tên nhóm hàng' và tính tổng doanh thu và lượng bán
df_summary = df.groupby('Tên nhóm hàng').agg({
    'Thành tiền': 'sum',
    'SL': 'sum'
}).reset_index()

# Đổi tên cột cho dễ hiểu
df_summary = df_summary.rename(columns={
    'Thành tiền': 'Doanh thu',
    'SL': 'Lượng bán'
})

df_summary

Tên nhóm hàng   Doanh thu  Lượng bán
0              Bột   626000000      15650
1          Set trà   777947000       8483
2  Trà củ, quả sấy   800280000      22594
3          Trà hoa  1877505000      22743
4          Trà mix   697342000      22876

##1.4. Dashboard tổng quan: Doanh thu trung bình theo ngày trong tháng

Doanh thu trung bình hàng ngày dao động khoảng từ 10M đến gần 14M VND.Có sự biến động theo từng ngày, có ngày doanh thu đột biến cao hơn rõ rệt.Có thể xem xét các yếu tố ngoại cảnh hoặc chương trình khuyến mãi để giải thích biến động này.

In [37]:
# Đảm bảo cột datetime
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

# Tạo cột 'Tháng' và 'Ngày trong tháng'
df['Tháng'] = df['Thời gian tạo đơn'].dt.month
df['Ngày trong tháng'] = df['Thời gian tạo đơn'].dt.day

# Bước 1: Tính tổng doanh thu và lượng bán theo từng (Tháng, Ngày trong tháng)
monthly_daily_total = df.groupby(['Tháng', 'Ngày trong tháng']).agg({
    'Thành tiền': 'sum',
    'SL': 'sum'
}).reset_index()

# Bước 2: Tính trung bình doanh thu và lượng bán cho từng 'Ngày trong tháng' qua các tháng
avg_by_day = monthly_daily_total.groupby('Ngày trong tháng').agg({
    'Thành tiền': 'mean',
    'SL': 'mean'
}).reset_index().rename(columns={
    'Thành tiền': 'Doanh thu trung bình',
    'SL': 'Lượng bán trung bình'
})

# Chuyển doanh thu trung bình sang hàng triệu và làm tròn 1 chữ số thập phân
avg_by_day['Doanh thu trung bình (triệu)'] = (avg_by_day['Doanh thu trung bình'] / 1_000_000).round(1)

# Làm tròn lượng bán trung bình tới số nguyên
avg_by_day['Lượng bán trung bình (đơn vị)'] = avg_by_day['Lượng bán trung bình'].round(0).astype(int)

# Chọn cột để hiển thị
result = avg_by_day[['Ngày trong tháng', 'Doanh thu trung bình (triệu)', 'Lượng bán trung bình (đơn vị)']]

result

Ngày trong tháng  Doanh thu trung bình (triệu)  \
0                  1                          12.7   
1                  2                          13.9   
2                  3                          13.7   
3                  4                          13.5   
4                  5                          13.1   
5                  6                          13.2   
6                  7                          13.7   
7                  8                          13.3   
8                  9                          13.4   
9                 10                          13.3   
10                11                          12.8   
11                12                          13.2   
12                13                          14.0   
13                14                          13.1   
14                15                          13.2   
15                16                          13.5   
16                17                          14.3   
17                18                          13.1   
18                19                          12.6   
19                20                          13.5   
20                21                          12.3   
21                22                          12.5   
22                23                          13.0   
23                24                          13.0   
24                25                          13.4   
25                26                          12.1   
26                27                          12.3   
27                28                          12.4   
28                29                          12.7   
29                30                          12.6   
30                31                          11.7   

    Lượng bán trung bình (đơn vị)  
0                             247  
1                             264  
2                             264  
3                             259  
4                             251  
5                             258  
6                             258  
7                             257  
8                             260  
9                             258  
10                            250  
11                            258  
12                            274  
13                            255  
14                            252  
15                            268  
16                            272  
17                            249  
18                            247  
19                            261  
20                            242  
21                            243  
22                            249  
23                            253  
24                            257  
25                            236  
26                            238  
27                            245  
28                            244  
29                            240  
30                            220

##1.5. Dashboard tổng quan: Xác suất bán hàng theo nhóm hàng

Nhóm hàng có xác suất bán hàng cao nhất là Trà mix (54.5%), tiếp theo là Trà hoa (54.4%), Trà củ, quả sấy (53.3%).Mặc dù Trà mix có xác suất bán hàng cao nhưng doanh thu lại thấp hơn Trà hoa và Trà củ, quả sấy.Đây có thể là tín hiệu cho thấy trà mix được bán phổ biến hơn, nhưng giá trị đơn hàng thấp hơn hoặc bán với giá thấp hơn.

In [38]:
# Tính số đơn hàng duy nhất theo nhóm hàng
don_hang_theo_nhom = df.groupby('Mã nhóm hàng')['Mã đơn hàng'].nunique()

# Tổng số đơn hàng duy nhất toàn bộ
tong_don_hang = df['Mã đơn hàng'].nunique()

# Tính xác suất bán hàng theo nhóm hàng nhân 100 (phần trăm), làm tròn 3 chữ số thập phân
xac_suat_ban_hang = ((don_hang_theo_nhom / tong_don_hang) * 100).round(1)

# Tạo DataFrame kết quả
df_xac_suat = xac_suat_ban_hang.reset_index().rename(columns={
    'Mã nhóm hàng': 'Nhóm hàng',
    'Mã đơn hàng': 'Xác suất bán hàng (%)'
})

df_xac_suat

Nhóm hàng  Xác suất bán hàng (%)
0       BOT                   40.3
1       SET                   23.9
2       THO                   54.4
3       TMX                   54.5
4       TTC                   53.3

##1.6. Dashboard tổng quan: Lượng bán trung bình theo từng ngày trong tháng

Lượng bán hàng và số lượng khách hàng khá ổn định trong suốt tháng.Tuy nhiên có sự sụt giảm đáng chú ý vào cuối tháng (ngày 30-31), có thể do yếu tố cuối tháng hoặc ngày nghỉ lễ.Nên phân tích thêm để hiểu nguyên nhân và có phương án thúc đẩy bán hàng cuối kỳ.


In [39]:
# Đảm bảo cột 'Thời gian tạo đơn' là datetime
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

# Tạo cột 'Ngày trong tháng'
df['Ngày trong tháng'] = df['Thời gian tạo đơn'].dt.day

# Tính tổng lượng bán (SL) và số đơn hàng (Mã đơn hàng) duy nhất theo từng ngày thực tế (date)
daily_agg = df.groupby(df['Thời gian tạo đơn'].dt.date).agg({
    'SL': 'sum',
    'Mã đơn hàng': 'nunique'  # số lượng đơn hàng duy nhất mỗi ngày
}).reset_index().rename(columns={'Thời gian tạo đơn': 'Ngày'})

# Tính trung bình lượng bán theo ngày trong tháng
avg_luong_ban = daily_agg.groupby(daily_agg['Ngày'].apply(lambda x: x.day)).agg({
    'SL': 'mean'
}).rename(columns={'SL': 'Lượng bán trung bình'}).reset_index()

# Tính tổng số đơn hàng (khách hàng) theo ngày trong tháng
tong_khach_hang = daily_agg.groupby(daily_agg['Ngày'].apply(lambda x: x.day)).agg({
    'Mã đơn hàng': 'sum'
}).rename(columns={'Mã đơn hàng': 'Số lượng khách hàng'}).reset_index()

# Gộp 2 bảng lại theo 'Ngày trong tháng'
result = pd.merge(avg_luong_ban, tong_khach_hang, left_on='Ngày', right_on='Ngày')

# Đổi tên cột 'Ngày' thành 'Ngày trong tháng'
result = result.rename(columns={'Ngày': 'Ngày trong tháng'})

result

Ngày trong tháng  Lượng bán trung bình  Số lượng khách hàng
0                  1            247.166667                 1028
1                  2            264.250000                 1068
2                  3            264.500000                 1096
3                  4            258.833333                 1064
4                  5            251.166667                 1028
5                  6            257.500000                 1054
6                  7            257.833333                 1055
7                  8            256.666667                 1044
8                  9            260.000000                 1064
9                 10            258.083333                 1075
10                11            249.583333                 1019
11                12            257.750000                 1056
12                13            274.166667                 1122
13                14            255.083333                 1042
14                15            252.166667                 1039
15                16            267.500000                 1062
16                17            271.833333                 1100
17                18            248.666667                 1031
18                19            247.416667                 1003
19                20            260.833333                 1040
20                21            242.083333                  990
21                22            243.333333                  993
22                23            249.083333                 1023
23                24            252.916667                 1027
24                25            257.416667                 1044
25                26            236.000000                  977
26                27            237.583333                  959
27                28            244.666667                  982
28                29            243.818182                  914
29                30            239.727273                  900
30                31            219.714286                  539

#II. PHÂN TÍCH HÀNG HÓA

##2.1. Dashboard phân tích hàng hoá: Pareto chart theo mặt hàng và nhóm hàng

Text luận giải:

Biểu đồ Pareto là công cụ giúp xác định các yếu tố chính ảnh hưởng đến doanh thu. Khi áp dụng biểu đồ này theo nhóm hàng và mặt hàng, ta dễ dàng phát hiện được rằng:

Theo nhóm hàng, nhóm [THO] Trà hoa đóng góp đến 39.3% tổng doanh thu, vượt xa các nhóm khác, cho thấy đây là nhóm sản phẩm mũi nhọn.

Các nhóm tiếp theo như [TTC] Trà củ, quả sấy, [SET] Set trà, và [TMX] Trà mix lần lượt góp phần tạo ra khoảng 87% doanh thu, cho thấy chỉ cần tập trung vào 4 nhóm là đã nắm bắt được phần lớn doanh số.

Theo mặt hàng, [BOT01] Bột cần tây dẫn đầu, chiếm khoảng 13.1%, tiếp theo là [TH006] Trà nhụy hoa nghệ tây (25.5%), [TTC01] Trà gừng (34.9%)... tạo thành một chuỗi "ít nhưng chất" – đúng theo nguyên lý 80/20.

→ Đề xuất: Doanh nghiệp nên tập trung nguồn lực vào các nhóm/mặt hàng thuộc top đầu của biểu đồ Pareto để tối ưu hóa nguồn lực và lợi nhuận, đồng thời có thể cân nhắc loại bỏ hoặc cải tiến những mặt hàng "đuôi dài" ít đóng góp vào doanh thu.

In [13]:
import matplotlib.pyplot as plt
from IPython.display import display

# Đọc dữ liệu
df = pd.read_csv('/content/drive/MyDrive/File/data_ggsheet.csv')

# Tính tổng doanh thu theo Nhóm hàng
group_rev = df.groupby('Mã nhóm hàng')['Thành tiền'].sum().reset_index()
group_rev = group_rev.sort_values(by='Thành tiền', ascending=False)

# Thêm cột % và % cộng dồn
group_rev['Tỷ lệ (%)'] = (group_rev['Thành tiền'] / group_rev['Thành tiền'].sum()) * 100
group_rev['% cộng dồn'] = group_rev['Tỷ lệ (%)'].cumsum()

# Hiển thị bảng Pareto theo Nhóm hàng
print("Bảng Pareto doanh thu theo nhóm hàng:")
display(group_rev)

# Tính tổng doanh thu theo Mặt hàng
item_rev = df.groupby('Tên mặt hàng')['Thành tiền'].sum().reset_index()
item_rev = item_rev.sort_values(by='Thành tiền', ascending=False)

# Thêm cột % và % cộng dồn
item_rev['Tỷ lệ (%)'] = (item_rev['Thành tiền'] / item_rev['Thành tiền'].sum()) * 100
item_rev['% cộng dồn'] = item_rev['Tỷ lệ (%)'].cumsum()

# Hiển thị bảng Pareto theo Mặt hàng
print("Bảng Pareto doanh thu theo mặt hàng:")
display(item_rev)


Bảng Pareto doanh thu theo nhóm hàng:


Mã nhóm hàng  Thành tiền  Tỷ lệ (%)  % cộng dồn
2          THO  1877505000  39.285958   39.285958
4          TTC   800280000  16.745503   56.031461
1          SET   777947000  16.278195   72.309657
3          TMX   697342000  14.591572   86.901228
0          BOT   626000000  13.098772  100.000000

Bảng Pareto doanh thu theo mặt hàng:


Tên mặt hàng  Thành tiền  Tỷ lệ (%)  % cộng dồn
0                        Bột cần tây   626000000  13.098772   13.098772
16             Trà nhuỵ hoa nghệ tây   592380000  12.395288   25.494060
12                          Trà gừng   432900000   9.058240   34.552300
14                 Trà hoa cúc trắng   421440000   8.818445   43.370745
1                            Cam lát   367380000   7.687263   51.058008
10                    Trà dưỡng nhan   317170000   6.636641   57.694650
18             Trà nụ hoa nhài trắng   290030000   6.068749   63.763399
15                  Trà hoa đậu biếc   279140000   5.840880   69.604279
9                     Trà cam sả quế   249888000   5.228795   74.833074
6       Set 10 gói trà hoa cúc trắng   243900000   5.103499   79.936573
7        Set 10 gói trà hoa đậu biếc   158712000   3.320978   83.257552
17          Trà nụ hoa hồng Tây Tạng   150315000   3.145275   86.402826
13                     Trà hoa Atiso   144200000   3.017321   89.420147
11                  Trà gạo lứt 8 vị   130284000   2.726135   92.146282
8   Set 10 gói trà nụ hoa nhài trắng   114600000   2.397954   94.544236
3          Set 10 gói trà dưỡng nhan    98475000   2.060546   96.604782
5                Set 10 gói trà gừng    80000000   1.673964   98.278746
2          Set 10 gói trà cam sả quế    50410000   1.054807   99.333553
4        Set 10 gói trà gạo lứt 8 vị    31850000   0.666447  100.000000

##2.2. Dashboard phân tích hàng hoá: Heatmap chart thể hiện doanh số của từng Phân khúc khách hàng mang lại theo từng mặt hàng và nhóm hàng

Text luận giải:

Việc phân tích doanh thu theo phân khúc khách hàng giúp chúng ta hiểu rõ đối tượng nào đang đóng góp lớn nhất vào từng dòng sản phẩm và từ đó có chiến lược marketing phù hợp.

Theo nhóm hàng:

*   Phân khúc A1 (khách hàng trưởng thành, thu nhập cao) là nhóm chi tiêu mạnh nhất ở hầu hết các nhóm hàng, đặc biệt là nhóm [THO] Trà hoa với doanh thu 512.8 triệu đồng.
*   Nhóm B1 và B2 cũng đóng góp đáng kể, nhất là ở nhóm [TMX] Trà mix và [TTC] Trà củ, quả sấy. Ngược lại, nhóm C2 và C3 đóng góp rất thấp, phản ánh sức mua yếu của phân khúc này.

Theo mặt hàng:

*   Mặt hàng Bột cần tây, Trà nhụy hoa nghệ tây, và Trà gừng là những sản phẩm phổ biến nhất ở nhiều phân khúc như A1, A2, B2.
*   Trà nhụy hoa nghệ tây ghi nhận mức doanh thu rất cao ở phân khúc A1, cho thấy đây là sản phẩm cao cấp có sức hút mạnh trong nhóm khách hàng thu nhập cao.
*   Mặt hàng thuộc nhóm SET (combo 10 gói trà các loại) thường có doanh thu thấp hơn và phân bổ đều, phù hợp nhóm C và sinh viên.


In [21]:
# Tạo bảng heatmap doanh thu theo nhóm hàng & phân khúc khách hàng
group_heatmap = df.groupby(['Tên nhóm hàng', 'Mã PKKH'])['Thành tiền'].sum().unstack(fill_value=0)

# Tạo bảng heatmap doanh thu theo mặt hàng & phân khúc khách hàng
item_heatmap = df.groupby(['Tên mặt hàng', 'Mã PKKH'])['Thành tiền'].sum().unstack(fill_value=0)

# Hiển thị kết quả dạng bảng
print("Doanh thu theo Tên nhóm hàng và Phân khúc KH:")
display(group_heatmap)

print("\nDoanh thu theo Tên mặt hàng và Phân khúc KH:")
display(item_heatmap)

Doanh thu theo Tên nhóm hàng và Phân khúc KH:


Mã PKKH                 A1         A2        A3         B1         B2  \
Tên nhóm hàng                                                           
Bột              139200000  137200000  59320000   72840000  103440000   
Set trà          154101000  146655000  63722000   93257000   79467000   
Trà củ, quả sấy  178545000  119595000  51510000  169035000  120225000   
Trà hoa          512778000  315284000  93680000  416084000  275389000   
Trà mix          157457000  110264000  45513000  154482000   96258000   

Mã PKKH                B3         C1        C2        C3  
Tên nhóm hàng                                             
Bột              44280000   15320000  34560000  19840000  
Set trà          47282000  155730000  24668000  13065000  
Trà củ, quả sấy  62370000   22320000  44010000  32670000  
Trà hoa          90816000   43996000  72413000  57065000  
Trà mix          43743000   21522000  38062000  30041000


Doanh thu theo Tên mặt hàng và Phân khúc KH:


Mã PKKH                                  A1         A2        A3         B1  \
Tên mặt hàng                                                                  
Bột cần tây                       139200000  137200000  59320000   72840000   
Cam lát                            89595000   40725000  40320000   56565000   
Set 10 gói trà cam sả quế          10153000    4757000  10295000    4047000   
Set 10 gói trà dưỡng nhan          23855000   13650000   4225000   13390000   
Set 10 gói trà gạo lứt 8 vị         6055000    2380000   6370000    5390000   
Set 10 gói trà gừng                22450000   14400000   2950000   10700000   
Set 10 gói trà hoa cúc trắng       41100000   65100000  11250000   32250000   
Set 10 gói trà hoa đậu biếc        28288000   28968000  15912000   14280000   
Set 10 gói trà nụ hoa nhài trắng   22200000   17400000  12720000   13200000   
Trà cam sả quế                     73986000   44726000  22078000   45524000   
Trà dưỡng nhan                     60935000   34020000   8855000   91840000   
Trà gạo lứt 8 vị                   22536000   31518000  14580000   17118000   
Trà gừng                           88950000   78870000  11190000  112470000   
Trà hoa Atiso                      25600000   10400000  13900000   37750000   
Trà hoa cúc trắng                  60560000  108480000  14640000   98080000   
Trà hoa đậu biếc                   70108000   53244000  23800000   51544000   
Trà nhuỵ hoa nghệ tây             279000000   69480000  11520000  136800000   
Trà nụ hoa hồng Tây Tạng           28435000   21615000  14740000   27170000   
Trà nụ hoa nhài trắng              49075000   52065000  15080000   64740000   

Mã PKKH                                  B2        B3        C1        C2  \
Tên mặt hàng                                                                
Bột cần tây                       103440000  44280000  15320000  34560000   
Cam lát                            72225000  31320000  12330000  17280000   
Set 10 gói trà cam sả quế           6958000   7384000   5822000    781000   
Set 10 gói trà dưỡng nhan          12090000   1820000  22555000   3770000   
Set 10 gói trà gạo lứt 8 vị         5145000   1820000   3255000    945000   
Set 10 gói trà gừng                 7200000   5150000  11650000   3350000   
Set 10 gói trà hoa cúc trắng       16050000  11700000  52800000   8550000   
Set 10 gói trà hoa đậu biếc        18224000   9928000  37808000   3672000   
Set 10 gói trà nụ hoa nhài trắng   13800000   9480000  21840000   3600000   
Trà cam sả quế                     19076000  26562000   6802000   7334000   
Trà dưỡng nhan                     52990000   9765000  10850000  25130000   
Trà gạo lứt 8 vị                   24192000   7416000   3870000   5598000   
Trà gừng                           48000000  31050000   9990000  26730000   
Trà hoa Atiso                      28150000   9350000   8300000   5000000   
Trà hoa cúc trắng                  58000000  25680000   7840000  24000000   
Trà hoa đậu biếc                   36924000  12716000   9316000  16728000   
Trà nhuỵ hoa nghệ tây              64080000  12420000   5040000   5940000   
Trà nụ hoa hồng Tây Tạng           29865000  10175000   7260000   7095000   
Trà nụ hoa nhài trắng              58370000  20475000   6240000  13650000   

Mã PKKH                                 C3  
Tên mặt hàng                                
Bột cần tây                       19840000  
Cam lát                            7020000  
Set 10 gói trà cam sả quế           213000  
Set 10 gói trà dưỡng nhan          3120000  
Set 10 gói trà gạo lứt 8 vị         490000  
Set 10 gói trà gừng                2150000  
Set 10 gói trà hoa cúc trắng       5100000  
Set 10 gói trà hoa đậu biếc        1632000  
Set 10 gói trà nụ hoa nhài trắng    360000  
Trà cam sả quế                     3800000  
Trà dưỡng nhan                    22785000  
Trà gạo lứt 8 vị                   3456000  
Trà gừng                          25650000  
Trà hoa Atiso                      5750000  
Tr

##2.3. Dashboard phân tích hàng hoá: Line chart thể hiện biến động doanh thu của các mặt hàng trong từng tháng được lọc riêng theo Phân khúc khách hàng A1

Biểu đồ thể hiện xu hướng thay đổi doanh thu theo từng tháng của từng mặt hàng riêng cho phân khúc khách hàng **A1**, cho thấy rõ hành vi mua hàng theo mùa vụ.

- Từ tháng 4 đến tháng 7, phần lớn các mặt hàng có doanh thu thấp. Điều này có thể phản ánh nhu cầu thấp trong giai đoạn giao mùa hoặc giữa năm.
- Tuy nhiên, bắt đầu từ tháng 8, các mặt hàng như **Trà nhụy hoa nghệ tây**, **Bột cần tây**, và **Trà gừng** ghi nhận sự tăng trưởng đột biến, đặc biệt là vào các tháng cuối năm (Tháng 10–12).
- Doanh thu tháng 12 đạt đỉnh với hầu hết mặt hàng, điều này có thể liên quan đến dịp lễ Tết, xu hướng mua biếu tặng hoặc trữ hàng.

Đề xuất:
- Tập trung thúc đẩy marketing và khuyến mãi vào tháng 8 trở đi – khi nhu cầu bắt đầu tăng lên mạnh mẽ.
- Xem xét lý do doanh thu thấp trong các tháng giữa năm (tháng 4–7) để có thể áp dụng gói ưu đãi hoặc sản phẩm thay thế phù hợp cho A1.


In [25]:
# Chuyển đổi cột 'Thời gian tạo đơn' thành datetime và lấy tháng
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])
df['Tháng'] = df['Thời gian tạo đơn'].dt.month

# Lọc dữ liệu cho phân khúc khách hàng A1
df_A1 = df[df['Mã PKKH'] == 'A1']

# Tính tổng doanh thu theo Tháng và Tên mặt hàng
df_a1_month_item = df_A1.groupby(['Tháng', 'Tên mặt hàng'])['Thành tiền'].sum().unstack(fill_value=0)

# Hiển thị bảng kết quả
display(df_a1_month_item)

Tên mặt hàng  Bột cần tây   Cam lát  Set 10 gói trà cam sả quế  \
Tháng                                                            
1                 8000000   5355000                     710000   
2                 7200000   4680000                     710000   
3                 6360000   3825000                     568000   
4                 2520000   1845000                     213000   
5                 3080000   1890000                     284000   
6                 3040000   2025000                     355000   
7                 3480000   1935000                     355000   
8                16080000   8595000                    1207000   
9                17600000  11115000                    1065000   
10               21200000  13455000                    1349000   
11               27440000  17370000                    1349000   
12               23200000  17505000                    1988000   

Tên mặt hàng  Set 10 gói trà dưỡng nhan  Set 10 gói trà gạo lứt 8 vị  \
Tháng                                                                  
1                               2015000                       455000   
2                               1755000                       350000   
3                               1560000                       385000   
4                                650000                       175000   
5                                845000                       175000   
6                                585000                       210000   
7                                780000                        70000   
8                               3055000                       910000   
9                               2405000                       665000   
10                              2470000                       840000   
11                              3120000                       700000   
12                              4615000                      1120000   

Tên mặt hàng  Set 10 gói trà gừng  Set 10 gói trà hoa cúc trắng  \
Tháng                                                             
1                         1950000                       3300000   
2                         1400000                       3000000   
3                         1550000                       2550000   
4                          650000                        600000   
5                          600000                       1500000   
6                          700000                       1350000   
7                          650000                       1200000   
8                         2850000                       4650000   
9                         2300000                       4950000   
10                        2500000                       4800000   
11                        2850000                       4650000   
12                        4450000                       8550000   

Tên mặt hàng  Set 10 gói trà hoa đậu biếc  Set 10 gói trà nụ hoa nhài trắng  \
Tháng                                                                         
1                                 2312000                           1440000   
2                                 1768000                           1200000   
3                                 1768000                           1320000   
4                                  680000                            720000   
5                                 1088000                            720000   
6                                  816000                            720000   
7                                  816000                            720000   
8                                 3536000                           3240000   
9                                 2992000                           2640000   
10                                3672000                           2520000   
11                                3264000                           2880000   
12                                5576000                           4080000   

Tên mặt hàng  Trà ca

##2.4. Dashboard phân tích hàng hoá: Số lượng khách hàng quay lại mua theo từng nhóm hàng và mặt hàng

Qua việc tổng hợp dữ liệu, ta thấy rằng các mặt hàng như Bột cần tây, Trà gừng và Cam lát ghi nhận số lượng khách hàng quay lại cao nhất, lần lượt đạt 645, 598 và 429 lượt. Đây là những sản phẩm có độ hấp dẫn và mức độ trung thành cao, có thể là nhờ vào hiệu quả sản phẩm hoặc giá trị trải nghiệm tích cực mà khách hàng nhận được.

Xét theo nhóm hàng, ba nhóm có số lượng khách hàng quay lại cao nhất là:

Trà hoa (THO): 13,745 lượt

Trà mix (TMX): 10,553 lượt

Trà củ, quả sấy (TTC): 8,446 lượt

Điều này chứng tỏ những nhóm sản phẩm này không chỉ thu hút được khách hàng mới mà còn giữ chân tốt khách hàng hiện tại, thể hiện sức hút dài hạn và tiềm năng duy trì doanh thu ổn định.

Ngược lại, nhóm Bột (BOT) và Set trà (SET) có lượng khách hàng quay lại thấp hơn đáng kể, cho thấy cần được xem xét lại về mặt chiến lược bán hàng, marketing hoặc chất lượng sản phẩm.

Đề xuất:

Tập trung phát triển chương trình chăm sóc khách hàng hoặc chiến dịch upsell, cross-sell dành riêng cho nhóm Trà hoa, Trà mix, TTC để tối ưu hóa CLV.

Tiến hành khảo sát chất lượng và nhu cầu thực tế với nhóm Bột và Set trà để xác định nguyên nhân và hướng cải thiện.

In [43]:
import itertools
from IPython.display import display

# Nhóm theo Khách hàng và đếm số lượng nhóm hàng đã từng mua
df_customer_group = df.groupby(['Tên khách hàng'])['Tên nhóm hàng'].apply(lambda x: list(set(x)))
df_customer_product = df.groupby(['Tên khách hàng'])['Tên mặt hàng'].apply(lambda x: list(set(x)))

# Đếm số lượng khách hàng quay lại mua theo nhóm hàng
group_counter = Counter(itertools.chain.from_iterable(df_customer_group))
df_return_by_group = pd.DataFrame(group_counter.items(), columns=['Nhóm hàng', 'Số lượng khách quay lại'])

# Đếm số lượng khách hàng quay lại mua theo mặt hàng
product_counter = Counter(itertools.chain.from_iterable(df_customer_product))
df_return_by_product = pd.DataFrame(product_counter.items(), columns=['Mặt hàng', 'Số lượng khách quay lại'])

display(df_return_by_product)

Mặt hàng  Số lượng khách quay lại
0                        Bột cần tây                     5036
1                      Trà hoa Atiso                     1975
2                            Cam lát                     3685
3                     Trà cam sả quế                     3231
4                     Trà dưỡng nhan                     3874
5                   Trà gạo lứt 8 vị                     3445
6              Trà nụ hoa nhài trắng                     2639
7                   Trà hoa đậu biếc                     2496
8              Trà nhuỵ hoa nghệ tây                     1858
9        Set 10 gói trà hoa đậu biếc                      936
10                          Trà gừng                     4759
11                 Trà hoa cúc trắng                     2874
12       Set 10 gói trà gạo lứt 8 vị                      793
13         Set 10 gói trà dưỡng nhan                     1202
14          Trà nụ hoa hồng Tây Tạng                     1897
15               Set 10 gói trà gừng                     1247
16  Set 10 gói trà nụ hoa nhài trắng                      810
17      Set 10 gói trà hoa cúc trắng                     1224
18         Set 10 gói trà cam sả quế                      615

##2.5. Dashboard phân tích giỏ hàng: Heatmap chart thể hiện Tần suất mua kèm giữa các cặp sản phẩm

Dữ liệu cho thấy cặp sản phẩm được mua kèm nhiều nhất là Bột cần tây & Trà gừng với 4,056 lượt – đây có thể là một combo "chủ lực" phục vụ nhu cầu giải độc, thanh lọc và tăng đề kháng. Theo sau đó là các cặp như Trà gừng & Cam lát (2,324 lượt), Trà gạo lứt 8 vị & Trà gừng (1,968 lượt) hay Trà dưỡng nhan & Trà gừng (1,487 lượt), qua đó thấy rõ vai trò trung tâm của Trà gừng trong hành vi mua combo – giống như một sản phẩm "mỏ neo", thường được kết hợp với nhiều món khác. Bên cạnh đó, Bột cần tây cũng là một sản phẩm xuất hiện dày đặc trong nhiều tổ hợp, đặc biệt là với các loại trà thiên về sức khỏe như Trà cam sả quế, Trà dưỡng nhan, hay Trà gạo lứt 8 vị, cho thấy người tiêu dùng thường lựa chọn Bột cần tây như một sản phẩm chức năng chính, còn các loại trà đi kèm có thể đóng vai trò tăng hương vị hoặc bổ sung hiệu quả tổng thể.

Ngoài ra, có thể thấy các sản phẩm như Trà gừng, Cam lát, Trà hoa cúc và Trà dưỡng nhan có tính “đa năng” hoặc “linh hoạt”, vì được kết hợp nhiều lần với các sản phẩm khác nhau. Điều này rất lý tưởng để triển khai các chương trình combo khuyến mãi thông minh, bundle gợi ý tự động hay chiến dịch upsell chuyên biệt.

Về mặt khai thác thực tế, có thể tập trung xây dựng nhóm combo ưu tiên dựa trên các cặp có lượt mua từ 1000 trở lên như “Bột cần tây & Trà gừng” hay “Cam lát & Trà gừng”. Đồng thời, khi người dùng thêm một trong hai sản phẩm của các combo phổ biến vào giỏ hàng, hệ thống nên tự động gợi ý thêm món còn lại để tăng tỷ lệ đơn hàng combo. Đặc biệt, các dịp nhạy cảm về sức khỏe như mùa lạnh, mùa dịch hay giai đoạn tăng cường miễn dịch sẽ là thời điểm phù hợp để đẩy mạnh marketing cho các combo có yếu tố Trà gừng, tận dụng tâm lý phòng bệnh của người tiêu dùng.

In [14]:
from itertools import combinations
from collections import Counter
grouped = df.groupby('Mã đơn hàng')['Tên mặt hàng'].apply(list)
display(grouped)
pair_counter = Counter()
for products in grouped:
    products = sorted(set(products))  # loại bỏ trùng sản phẩm trong 1 đơn
    for pair in combinations(products, 2):
        pair_counter[pair] += 1

df_pairs = pd.DataFrame(pair_counter.items(), columns=['Product Pair', 'Số lần mua kèm'])
df_pairs[['Sản phẩm A', 'Sản phẩm B']] = pd.DataFrame(df_pairs['Product Pair'].tolist(), index=df_pairs.index)
df_pairs.drop(columns='Product Pair', inplace=True)

#df_pairs = df_pairs.sort_values(by="Số lần mua kèm", ascending=False)
display(df_pairs)
#output_path = '/content/drive/MyDrive/File/product_pair_combined.csv'
#df_pairs.to_csv(output_path, index=False, encoding='utf-8-sig')

Mã đơn hàng
ORD0000001    [Bột cần tây, Set 10 gói trà hoa cúc trắng, Tr...
ORD0000002    [Bột cần tây, Set 10 gói trà hoa cúc trắng, Tr...
ORD0000003                  [Trà dưỡng nhan, Trà gừng, Cam lát]
ORD0000004                          [Trà gạo lứt 8 vị, Cam lát]
ORD0000005           [Trà hoa cúc trắng, Trà nhuỵ hoa nghệ tây]
                                    ...                        
ORD0031434                        [Bột cần tây, Trà dưỡng nhan]
ORD0031435                     [Bột cần tây, Trà gừng, Cam lát]
ORD0031436                      [Bột cần tây, Trà gạo lứt 8 vị]
ORD0031437                        [Bột cần tây, Trà cam sả quế]
ORD0031438                               [Bột cần tây, Cam lát]
Name: Tên mặt hàng, Length: 31438, dtype: object

Số lần mua kèm                    Sản phẩm A  \
0               484                   Bột cần tây   
1              4056                   Bột cần tây   
2               865                   Bột cần tây   
3               428  Set 10 gói trà hoa cúc trắng   
4                99  Set 10 gói trà hoa cúc trắng   
..              ...                           ...   
166              23     Set 10 gói trà cam sả quế   
167              36   Set 10 gói trà gạo lứt 8 vị   
168              40     Set 10 gói trà cam sả quế   
169              20     Set 10 gói trà dưỡng nhan   
170              31     Set 10 gói trà cam sả quế   

                           Sản phẩm B  
0        Set 10 gói trà hoa cúc trắng  
1                            Trà gừng  
2                       Trà hoa Atiso  
3                            Trà gừng  
4                       Trà hoa Atiso  
..                                ...  
166               Set 10 gói trà gừng  
167               Set 10 gói trà gừng  
168             Trà nhuỵ hoa nghệ tây  
169       Set 10 gói trà gạo lứt 8 vị  
170  Set 10 gói trà nụ hoa nhài trắng  

[171 rows x 3 columns]

##2.6. Dashboard phân tích giỏ hàng: Top cặp sản phẩm được mua cùng nhau nhiều nhất.

Chart này thể hiện Top 10 cặp sản phẩm được mua cùng nhau nhiều nhất, giúp làm rõ các hành vi mua hàng có tính kết hợp cao – đây là một trong những cơ sở rất quan trọng để xây dựng các gói combo khuyến mãi hoặc hệ thống gợi ý sản phẩm tự động trong quá trình bán hàng.

Có thể thấy rằng “Bột cần tây & Trà gừng” là cặp sản phẩm chiếm ưu thế tuyệt đối với 4.056 lượt mua kèm, vượt xa cặp đứng thứ hai là “Trà dưỡng nhan & Trà gừng” (2.676 lượt). Điều này cho thấy không chỉ Trà gừng là sản phẩm cốt lõi trong các hành vi mua hàng đi kèm, mà Bột cần tây cũng giữ vai trò là một sản phẩm chức năng chủ lực – thường được kết hợp với các loại trà để tạo nên sự cân bằng giữa “giải độc – thư giãn – tăng đề kháng”.

Ngoài ra, các combo như “Cam lát & Trà gừng”, “Bột cần tây & Cam lát”, hay “Bột cần tây & Trà hoa cúc trắng” cũng có lượt mua kèm cao, phản ánh xu hướng khách hàng lựa chọn combo hướng đến sức khỏe thể chất – cải thiện tiêu hóa – nâng cao hệ miễn dịch.

Từ biểu đồ này, có thể đề xuất triển khai một số hướng chiến lược:

Tạo các gói combo đề xuất sẵn tại điểm bán (website, app hoặc cửa hàng vật lý), ưu tiên các combo trong Top 5 vì có tỉ lệ chấp nhận cao.

Gợi ý mua kèm thông minh trên giao diện giỏ hàng hoặc trang sản phẩm – ví dụ: nếu khách chọn Trà gừng, hệ thống tự động gợi ý Bột cần tây hoặc Cam lát.

Thiết kế các chương trình khuyến mãi theo mùa, đặc biệt là vào mùa lạnh hoặc cao điểm dịch bệnh – ưu tiên các combo có yếu tố Trà gừng để tận dụng hành vi mua hàng phòng ngừa bệnh.

In [17]:
df_pairs = df_pairs.sort_values(by="Số lần mua kèm", ascending=False)
display(df_pairs)

Số lần mua kèm                 Sản phẩm A                    Sản phẩm B
1              4056                Bột cần tây                      Trà gừng
8              2676             Trà dưỡng nhan                      Trà gừng
36             2554                Bột cần tây                       Cam lát
14             2469                Bột cần tây                Trà dưỡng nhan
7              2324                    Cam lát                      Trà gừng
..              ...                        ...                           ...
145              25  Set 10 gói trà cam sả quế     Set 10 gói trà dưỡng nhan
166              23  Set 10 gói trà cam sả quế           Set 10 gói trà gừng
132              22  Set 10 gói trà cam sả quế  Set 10 gói trà hoa cúc trắng
169              20  Set 10 gói trà dưỡng nhan   Set 10 gói trà gạo lứt 8 vị
163              15  Set 10 gói trà cam sả quế   Set 10 gói trà hoa đậu biếc

[171 rows x 3 columns]

##2.7. Dashboard phân tích giỏ hàng: Tỷ lệ đơn hàng theo combo và đơn lẻ.

Biểu đồ tròn này thể hiện tỷ lệ giữa đơn hàng combo (mua từ 2 sản phẩm trở lên) và đơn hàng đơn lẻ (chỉ mua 1 sản phẩm). Dữ liệu cho thấy:

- 90.7% đơn hàng là combo, chiếm đại đa số.
- Chỉ có 9.3% đơn hàng là đơn lẻ.

Điều này phản ánh một xu hướng rất rõ ràng: phần lớn khách hàng có xu hướng mua nhiều sản phẩm trong cùng một đơn hàng. Có thể do nhu cầu phối hợp sử dụng các sản phẩm (đặc biệt là các loại trà & bột chức năng), hoặc do hiệu ứng của các chương trình khuyến mãi/upsell hiện tại đã thúc đẩy hành vi mua combo.
Từ đây, có một số luận điểm và đề xuất chiến lược quan trọng:
Thị trường phản ứng tốt với hành vi mua theo combo. Điều này mở ra cơ hội để:
+ Tối ưu hóa các gói sản phẩm theo chủ đề, ví dụ: “Giải độc – thanh lọc”, “Tăng đề kháng – làm đẹp”, v.v.
+ Tích hợp khuyến mãi theo combo: giảm giá nhẹ khi mua từ 2–3 sản phẩm cùng lúc, hoặc tặng kèm sản phẩm thứ 3.

Hành vi mua lặp lại khả năng cao xuất phát từ combo: Khách hàng có xu hướng quay lại mua nguyên bộ sản phẩm thay vì từng món riêng lẻ → cần phân tích kỹ hơn xem combo nào có tỉ lệ lặp lại cao nhất để lên kế hoạch chăm sóc khách hàng.

Gợi ý bán chéo nên tập trung vào combo thay vì sản phẩm đơn lẻ: Vì xác suất người dùng mua thêm cao hơn nếu thấy lợi ích rõ ràng về sự kết hợp, nên cần xây dựng logic gợi ý theo hành vi này trên website, app hoặc tại cửa hàng.

In [ ]:
# Đếm số lượng sản phẩm trong mỗi đơn hàng
df_order_count = df.groupby('Mã đơn hàng')['Tên mặt hàng'].nunique().reset_index()
df_order_count['Loại đơn'] = df_order_count['Tên mặt hàng'].apply(lambda x: 'Combo (≥2 sản phẩm)' if x >= 2 else 'Đơn lẻ (1 sản phẩm)')

# Tính tổng số đơn theo từng loại
order_type_summary = df_order_count['Loại đơn'].value_counts().reset_index()
order_type_summary.columns = ['Loại đơn hàng', 'Số lượng đơn hàng']
print(order_type_summary.to_string(index=False))
# Xuất file CSV
#order_type_summary.to_csv('/content/drive/MyDrive/File/order_combo_ratio.csv', index=False, encoding='utf-8-sig')

      Loại đơn hàng  Số lượng đơn hàng
Combo (≥2 sản phẩm)              28509
Đơn lẻ (1 sản phẩm)               2929


##2.8. Dashboard phân tích giỏ hàng: Biến động mua kèm theo tháng của các combo sản phẩm.

Biểu đồ trên cho thấy rõ ràng xu hướng tăng mạnh số lượt mua kèm đối với các combo sản phẩm trong những tháng cuối năm, đặc biệt từ tháng 8 đến tháng 11. Nổi bật nhất là combo “Bột cần tây & Trà gừng”, với số lượt mua tăng liên tục và đạt đỉnh hơn 900 lượt trong tháng 11, sau đó có xu hướng giảm nhẹ trong tháng 12. Điều này có thể phản ánh ảnh hưởng của yếu tố mùa vụ — thời tiết lạnh hơn khiến khách hàng có xu hướng chọn các sản phẩm tốt cho sức khỏe, thanh lọc và tăng đề kháng. Các combo như “Cam lát & Trà gừng”, “Trà cam sả quế & Trà gừng” hay “Bột cần tây & Cam lát” cũng ghi nhận mức tăng đáng kể trong cùng giai đoạn.

Dữ liệu cũng cho thấy các combo có yếu tố "thanh nhiệt – hỗ trợ tiêu hóa" hoặc "giải độc – tăng đề kháng" có khả năng được ưu tiên hơn trong quyết định mua kèm. Điều này mở ra cơ hội định hướng chiến lược marketing tập trung vào mùa thu và đông, đồng thời có thể tối ưu gợi ý combo trên nền tảng bán hàng.

In [11]:
# Chuẩn hóa định dạng thời gian thành tháng
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'], errors='coerce')
df['Tháng'] = df['Thời gian tạo đơn'].dt.strftime('%m')

# Gom mặt hàng theo từng đơn hàng + tháng
grouped = df.groupby(['Mã đơn hàng', 'Tháng'])['Tên mặt hàng'].apply(list).reset_index()

# Đếm các cặp sản phẩm theo từng tháng
monthly_pair_counter = []
for _, row in grouped.iterrows():
    products = sorted(set(row['Tên mặt hàng']))
    month = row['Tháng']
    for pair in combinations(products, 2):
        monthly_pair_counter.append((pair[0], pair[1], month))

# Tạo DataFrame tần suất theo tháng
df_monthly_pairs = pd.DataFrame(monthly_pair_counter, columns=['Sản phẩm A', 'Sản phẩm B', 'Tháng'])
df_monthly_pairs['Số lần mua kèm'] = 1

# Tổng hợp số lần mua kèm theo cặp sản phẩm và tháng
pair_by_month = df_monthly_pairs.groupby(['Sản phẩm A', 'Sản phẩm B', 'Tháng'])['Số lần mua kèm'].sum().reset_index()
display(pair_by_month)
#pair_by_month.to_csv('/content/drive/MyDrive/File/pair_by_month.csv', index=False, encoding='utf-8-sig')

Sản phẩm A             Sản phẩm B Tháng  Số lần mua kèm
0                  Bột cần tây                Cam lát    01             159
1                  Bột cần tây                Cam lát    02             107
2                  Bột cần tây                Cam lát    03              86
3                  Bột cần tây                Cam lát    04              68
4                  Bột cần tây                Cam lát    05              81
...                        ...                    ...   ...             ...
2015  Trà nụ hoa hồng Tây Tạng  Trà nụ hoa nhài trắng    08              38
2016  Trà nụ hoa hồng Tây Tạng  Trà nụ hoa nhài trắng    09              32
2017  Trà nụ hoa hồng Tây Tạng  Trà nụ hoa nhài trắng    10              24
2018  Trà nụ hoa hồng Tây Tạng  Trà nụ hoa nhài trắng    11              34
2019  Trà nụ hoa hồng Tây Tạng  Trà nụ hoa nhài trắng    12              32

[2020 rows x 4 columns]

#III. PHÂN TÍCH KHÁCH HÀNG

##3.1. Dashboard phân tích khách hàng: Row card thể hiện chỉ số tổng quan.

Về phần khách hàng, nhằm mong muốn phân tích xem được hành vi của khách hàng hay nói đúng hơn là khả năng quay lại của khách hàng, thì nhóm thực hiện đi sâu vào phân tích các khía cạnh xung quanh khách hàng quay lại mua sản phẩm. Đầu tiên với dữ liệu ta có được 26.575 khách hàng quay lại trong năm 2022 với tỷ lệ quay lại là 49,64% Đem lại 4071 triệu VND cho doanh thu của cửa hàng


In [26]:
orders_per_customer = df.groupby('Mã khách hàng')['Mã đơn hàng'].nunique().reset_index()
orders_per_customer.columns = ['Mã khách hàng', 'Số đơn hàng']
orders_per_customer['Phân loại'] = orders_per_customer['Số đơn hàng'].apply(
    lambda x: 'Khách quay lại' if x > 1 else 'Khách mới'
)

# Tính tỷ lệ khách hàng quay lại
returning_ratio = orders_per_customer['Phân loại'].value_counts(normalize=True) * 100
ty_le_quay_lai = returning_ratio.get('Khách quay lại', 0)

# Lọc khách quay lại
returning_customers = orders_per_customer[orders_per_customer['Phân loại'] == 'Khách quay lại']['Mã khách hàng']
df_returning = df[df['Mã khách hàng'].isin(returning_customers)]

# Tính doanh thu và số đơn hàng của khách quay lại
doanh_thu = df_returning['Thành tiền'].sum()
so_don = df_returning['Mã đơn hàng'].nunique()

# In ra kết quả
print("Tỷ lệ khách hàng quay lại:", round(ty_le_quay_lai, 2), "%")
print("Tổng doanh thu từ khách quay lại:", round(doanh_thu, 0))
print("Tổng số đơn hàng từ khách quay lại:", so_don)

Tỷ lệ khách hàng quay lại: 49.64 %
Tổng doanh thu từ khách quay lại: 4071469000
Tổng số đơn hàng từ khách quay lại: 26575


##3.2. Dashboard phân tích khách hàng: Pareto chart thể hiện số lượng khách hàng quay lại qua các tháng.

Tập trung vào các tháng cao điểm từ 8 đến 12 đóng góp khoảng 63,57% doanh thu. Ngoài ra thì lượng khách hàng quay lại cũng có xu hướng tăng dần qua từng tháng

Nhìn vào biến động của khách hàng quay lại theo tháng, ta có thể thấy được xu hướng của khách hàng khi họ bắt đầu quay lại mua hàng nhiều hơn ở các tháng cuối năm, bắt đầu từ tháng 7 trở đi thì lượng khách hàng tăng vọt từ 1300 khách hàng lên đến trên 2000 khách hàng và vẫn duy trì xu hướng tăng dần theo thời gian cho đến cuối tháng

In [27]:
# Chuyển cột 'Thời gian tạo đơn' thành định dạng datetime
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

# Trích xuất tháng từ cột 'Thời gian tạo đơn'
df['Tháng'] = df['Thời gian tạo đơn'].dt.month

# Tìm khách hàng quay lại (có nhiều hơn một đơn hàng)
customer_orders = df.groupby('Mã khách hàng')['Mã đơn hàng'].nunique()
returning_customers = customer_orders[customer_orders > 1].index

# Lọc dữ liệu chỉ giữ các đơn hàng từ khách hàng quay lại
df_returning = df[df['Mã khách hàng'].isin(returning_customers)]

# Đếm số lượng khách hàng quay lại duy nhất theo từng tháng
monthly_returning_customers = df_returning.groupby('Tháng')['Mã khách hàng'].nunique().reset_index()

# Tính tổng doanh thu của khách hàng quay lại theo từng tháng
monthly_revenue = df_returning.groupby('Tháng')['Thành tiền'].sum().reset_index()

# Gộp dữ liệu số lượng khách hàng quay lại và doanh thu
monthly_data = monthly_returning_customers.merge(monthly_revenue, on='Tháng', how='left')

# Đặt lại tên cột
monthly_data.columns = ['Tháng', 'Số lượng khách hàng quay lại', 'Doanh thu']

# Đảm bảo tất cả các tháng từ 1 đến 12 được hiển thị (điền 0 nếu không có)
all_months = pd.DataFrame({'Tháng': range(1, 13)})
monthly_data = all_months.merge(monthly_data, on='Tháng', how='left').fillna(0)

# Tính phần trăm doanh thu
total_revenue = monthly_data['Doanh thu'].sum()
monthly_data['Phần trăm doanh thu'] = (monthly_data['Doanh thu'] / total_revenue * 100).round(2)

# Chuyển cột 'Tháng' thành định dạng "Tháng 01", "Tháng 02",...
monthly_data['Tháng'] = monthly_data['Tháng'].apply(lambda x: f'Tháng {x:02d}')

# Chuyển cột 'Số lượng khách hàng quay lại' và 'Doanh thu' thành kiểu số nguyên
monthly_data['Số lượng khách hàng quay lại'] = monthly_data['Số lượng khách hàng quay lại'].astype(int)
monthly_data['Doanh thu'] = monthly_data['Doanh thu'].astype(int)

# Hiển thị DataFrame
print(monthly_data)

       Tháng  Số lượng khách hàng quay lại  Doanh thu  Phần trăm doanh thu
0   Tháng 01                           830  188853000                 4.64
1   Tháng 02                          1226  240483000                 5.91
2   Tháng 03                          1281  224454000                 5.51
3   Tháng 04                          1109  170942000                 4.20
4   Tháng 05                          1204  188052000                 4.62
5   Tháng 06                          1269  228571000                 5.61
6   Tháng 07                          1331  241699000                 5.94
7   Tháng 08                          2050  449136000                11.03
8   Tháng 09                          2360  468130000                11.50
9   Tháng 10                          2486  508063000                12.48
10  Tháng 11                          2486  564227000                13.86
11  Tháng 12                          2604  598859000                14.71


##3.3. Dashboard phân tích khách hàng: Doanh thu theo phân loại khách hàng là khách mới hay khách quay lại.

Trong đó nhóm hàng mà khách hàng quay lại mua nhiều nhất là nhóm trà hoa với 1636 triệu VND, chiếm đa số trong các nhóm hàng có mặt tại cửa hàng, với các mặt hàng được ưa chuộng như trà hoa cúc trăng, trà nụ hoa nhài và trà hoa đậu biếc đều trên 3k lượt bán

In [28]:
# Làm sạch và chuẩn hóa dữ liệu
df.columns = [col.strip() for col in df.columns]
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

# Tính số đơn hàng theo khách hàng
order_count = df.groupby('Mã khách hàng')['Mã đơn hàng'].nunique().reset_index()
order_count.columns = ['Mã khách hàng', 'Số đơn hàng']
order_count['Phân loại'] = order_count['Số đơn hàng'].apply(lambda x: 'Khách quay lại' if x > 1 else 'Khách mới')

# Lọc danh sách khách quay lại
returning_customers = order_count[order_count['Phân loại'] == 'Khách quay lại']['Mã khách hàng']

# Lọc dữ liệu giao dịch của khách quay lại
df_returning = df[df['Mã khách hàng'].isin(returning_customers)]

# Tính doanh thu (tổng Tổng tiền) theo nhóm hàng
doanh_thu_theo_nhom_hang = df_returning.groupby('Tên nhóm hàng')['Thành tiền'].sum().reset_index()

# Đặt lại tên cột
doanh_thu_theo_nhom_hang.columns = ['Tên nhóm hàng', 'Doanh thu']

# Sắp xếp giảm dần theo doanh thu
doanh_thu_theo_nhom_hang = doanh_thu_theo_nhom_hang.sort_values(by='Doanh thu', ascending=False)

# Chuyển cột 'Doanh thu' thành kiểu số nguyên (nếu cần)
doanh_thu_theo_nhom_hang['Doanh thu'] = doanh_thu_theo_nhom_hang['Doanh thu'].astype(int)

# Hiển thị DataFrame
print(doanh_thu_theo_nhom_hang)
# Đếm số đơn hàng duy nhất (distinct) theo 'Tên mặt hàng' từ khách quay lại
doanh_so_luot_ban = df_returning.groupby('Tên mặt hàng')['Mã đơn hàng'].nunique().reset_index()

# Đổi tên cột cho rõ
doanh_so_luot_ban.columns = ['Tên mặt hàng', 'Số lượt bán (khách quay lại)']

# Sắp xếp giảm dần theo số lượt bán
doanh_so_luot_ban = doanh_so_luot_ban.sort_values(by='Số lượt bán (khách quay lại)', ascending=False)

# Danh sách các mặt hàng cần lọc
mat_hang_can_loc = [
    'Trà hoa cúc trắng',
    'Trà nụ hoa nhài trắng',
    'Trà hoa đậu biếc'
]

# Lọc các mặt hàng này
ket_qua = doanh_so_luot_ban[doanh_so_luot_ban['Tên mặt hàng'].isin(mat_hang_can_loc)]

# Hiển thị kết quả
print(ket_qua)


     Tên nhóm hàng   Doanh thu
3          Trà hoa  1636165000
2  Trà củ, quả sấy   678810000
1          Set trà   629667000
4          Trà mix   595747000
0              Bột   531080000
             Tên mặt hàng  Số lượt bán (khách quay lại)
14      Trà hoa cúc trắng                          4184
18  Trà nụ hoa nhài trắng                          3629
15       Trà hoa đậu biếc                          3320


##3.4. Dashboard phân tích khách hàng: Doanh thu và số lượt sản phẩm bán ra

Tuy nhiên số lượt bán lớn ko đông nghĩa với doanh thu mang lại cao nhất, bất ngờ thì mặt hàng dẫn đầu về doanh thu của Trà hoa lại là trà nhụy hoa nghệ tây với 540 tr VND nhưng chỉ với 2813 lượt bán ra

In [30]:
# Làm sạch và chuẩn hóa dữ liệu
df.columns = [col.strip() for col in df.columns]
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

# Tính số đơn hàng theo khách hàng
order_count = df.groupby('Mã khách hàng')['Mã đơn hàng'].nunique().reset_index()
order_count.columns = ['Mã khách hàng', 'Số đơn hàng']
order_count['Phân loại'] = order_count['Số đơn hàng'].apply(lambda x: 'Khách quay lại' if x > 1 else 'Khách mới')

# Lọc danh sách khách quay lại
returning_customers = order_count[order_count['Phân loại'] == 'Khách quay lại']['Mã khách hàng']

# Lọc dữ liệu giao dịch của khách quay lại
df_returning = df[df['Mã khách hàng'].isin(returning_customers)]

# Tính tổng số lượng (SL) và doanh thu (Tổng tiền) theo Tên mặt hàng
doanh_thu_sl_theo_mat_hang = df_returning.groupby('Tên mặt hàng').agg({
    'SL': 'sum',
    'Thành tiền': 'sum'
}).reset_index()

# Đặt lại tên cột
doanh_thu_sl_theo_mat_hang.columns = ['Tên mặt hàng', 'Tổng số lượng (SL)', 'Doanh thu']

# Sắp xếp giảm dần theo Doanh thu
doanh_thu_sl_theo_mat_hang = doanh_thu_sl_theo_mat_hang.sort_values(by='Doanh thu', ascending=False)

# Chuyển cột 'Tổng số lượng (SL)' và 'Doanh thu' thành kiểu số nguyên (nếu cần)
doanh_thu_sl_theo_mat_hang['Tổng số lượng (SL)'] = doanh_thu_sl_theo_mat_hang['Tổng số lượng (SL)'].astype(int)
doanh_thu_sl_theo_mat_hang['Doanh thu'] = doanh_thu_sl_theo_mat_hang['Doanh thu'].astype(int)

# Hiển thị DataFrame
print(doanh_thu_sl_theo_mat_hang)

# Đếm số đơn hàng duy nhất (distinct) theo 'Tên mặt hàng' từ khách quay lại
doanh_so_luot_ban = df_returning.groupby('Tên mặt hàng')['Mã đơn hàng'].nunique().reset_index()

# Đổi tên cột cho rõ
doanh_so_luot_ban.columns = ['Tên mặt hàng', 'Số lượt bán (khách quay lại)']

# Sắp xếp giảm dần theo số lượt bán
doanh_so_luot_ban = doanh_so_luot_ban.sort_values(by='Số lượt bán (khách quay lại)', ascending=False)



# Lọc đúng tên mặt hàng "Trà nhuỵ hoa nghệ tây"
ket_qua = doanh_so_luot_ban[doanh_so_luot_ban['Tên mặt hàng'] == 'Trà nhuỵ hoa nghệ tây']

# Hiển thị kết quả
print(ket_qua)

                        Tên mặt hàng  Tổng số lượng (SL)  Doanh thu
16             Trà nhuỵ hoa nghệ tây                3000  540000000
0                        Bột cần tây               13277  531080000
12                          Trà gừng               12247  367410000
14                 Trà hoa cúc trắng                4488  359040000
1                            Cam lát                6920  311400000
10                    Trà dưỡng nhan                7669  268415000
18             Trà nụ hoa nhài trắng                3836  249340000
15                  Trà hoa đậu biếc                3515  239020000
9                     Trà cam sả quế                5689  216182000
6       Set 10 gói trà hoa cúc trắng                1321  198150000
17          Trà nụ hoa hồng Tây Tạng                2313  127215000
7        Set 10 gói trà hoa đậu biếc                 931  126616000
13                     Trà hoa Atiso                2431  121550000
11                  Trà gạo lứt 8 vị            

##3.5. Dashboard phân tích khách hàng: Doanh thu và số lượt sản phẩm bán ra

Giờ ta sẽ tiến hành đi sâu vào để tìm hiểu xu hướng của các khách hàng quay lại đặc biệt là từ tháng 8 trở đi với các con số tăng trưởng mạnh mẽ. Đầu tiên ta thấy được xu hướng quay trở lại mua hàng của khách hàng. Ta thấy được khách hàng quay trở lại mua hàng với các ngày trong tháng khá là đều nhau dao động với mức độ chênh lệch không quá lớn, tuy nhiên thì ngày 31 lại khách hàng ít quay trở lại mua hàng với bằng chứng là con số đó giảm xuống khoảng 200 khách so với ngày trước đó.

In [31]:
df.columns = [col.strip() for col in df.columns]
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

# Tính số đơn hàng theo khách hàng
order_count = df.groupby('Mã khách hàng')['Mã đơn hàng'].nunique().reset_index()
order_count.columns = ['Mã khách hàng', 'Số đơn hàng']
order_count['Phân loại'] = order_count['Số đơn hàng'].apply(lambda x: 'Khách quay lại' if x > 1 else 'Khách mới')

# Lấy danh sách khách quay lại
returning_customers = order_count[order_count['Phân loại'] == 'Khách quay lại']['Mã khách hàng']

# Lọc dữ liệu khách quay lại
df_returning = df[df['Mã khách hàng'].isin(returning_customers)]

# Lọc từ tháng 8 đến tháng 12
df_returning = df_returning[
    (df_returning['Thời gian tạo đơn'].dt.month >= 8) &
    (df_returning['Thời gian tạo đơn'].dt.month <= 12)
]

# Lấy ngày trong tháng
df_returning['Ngày trong tháng'] = df_returning['Thời gian tạo đơn'].dt.day

# Đếm số khách quay lại duy nhất theo ngày trong tháng (gộp nhiều tháng)
bien_dong_ngay_trong_thang = df_returning.groupby('Ngày trong tháng')['Mã khách hàng'].nunique().reset_index()
bien_dong_ngay_trong_thang.columns = ['Ngày trong tháng', 'Số khách quay lại']

# Sắp xếp tăng dần theo ngày
bien_dong_ngay_trong_thang = bien_dong_ngay_trong_thang.sort_values(by='Ngày trong tháng')

# Hiển thị kết quả
print(bien_dong_ngay_trong_thang)

    Ngày trong tháng  Số khách quay lại
0                  1                535
1                  2                533
2                  3                535
3                  4                529
4                  5                524
5                  6                497
6                  7                519
7                  8                507
8                  9                563
9                 10                526
10                11                496
11                12                523
12                13                541
13                14                508
14                15                508
15                16                524
16                17                537
17                18                512
18                19                499
19                20                508
20                21                494
21                22                472
22                23                487
23                24                503


Ta thấy được khách hàng quay trở lại mua hàng với các ngày trong tháng khá là đều nhau dao động với mức độ chênh lệch không quá lớn, tuy nhiên thì ngày 31 lại khách hàng ít quay trở lại mua hàng với bằng chứng là con số đó giảm xuống khoảng 200 khách so với ngày trước đó. Và khách hàng vào các ngày này cũng thực hiện mua sắm các sản phẩm ít ưa chuộng các sản phẩm về trà hoa khi ta thấy được họ có xu hướng ưa thích các loại trà dễ pha chế , tiện lợi. Nhìn vào biểu đồ thể hiện theo giờ của ngày 31 thì ta thấy được xu hướng mua hàng của khách hàng vào các ngày này cao điểm tại các khung giờ 10-11 giờ sáng, 18-19 và 21-22 giờ khuya, ta cần bố trí nhân lực phù hợp để tối ưu nhân lực.

In [33]:
import warnings

# Tắt cảnh báo SettingWithCopyWarning cho gọn
warnings.simplefilter(action='ignore', category=pd.errors.SettingWithCopyWarning)
df.columns = [col.strip() for col in df.columns]
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

# Tính số đơn hàng theo khách hàng
order_count = df.groupby('Mã khách hàng')['Mã đơn hàng'].nunique().reset_index()
order_count.columns = ['Mã khách hàng', 'Số đơn hàng']
order_count['Phân loại'] = order_count['Số đơn hàng'].apply(lambda x: 'Khách quay lại' if x > 1 else 'Khách mới')

# Lấy danh sách khách quay lại
returning_customers = order_count[order_count['Phân loại'] == 'Khách quay lại']['Mã khách hàng']

# Lọc dữ liệu khách quay lại
df_returning = df[df['Mã khách hàng'].isin(returning_customers)]

# Lọc tháng từ 8 đến 12
df_returning = df_returning[
    (df_returning['Thời gian tạo đơn'].dt.month >= 8) &
    (df_returning['Thời gian tạo đơn'].dt.month <= 12)
]

# Lọc chỉ lấy ngày 31
df_returning_31 = df_returning[df_returning['Thời gian tạo đơn'].dt.day == 31]

# Tạo cột "Giờ" (lấy giờ theo khung 8-23)
df_returning_31.loc[:, 'Giờ'] = df_returning_31['Thời gian tạo đơn'].dt.hour

# Lọc chỉ giữ các giờ từ 8 đến 23
df_returning_31 = df_returning_31[(df_returning_31['Giờ'] >= 8) & (df_returning_31['Giờ'] <= 23)]

# --- Phần 1: Hiển thị bảng số khách theo từng giờ, giờ dạng "8:00-8:59" ---
def format_gio(gio):
    return f"{gio}:00-{gio}:59"

khung_gio = df_returning_31.groupby('Giờ')['Mã khách hàng'].nunique().reset_index()
khung_gio.columns = ['Giờ', 'Số khách quay lại']
khung_gio['Giờ'] = khung_gio['Giờ'].apply(format_gio)
khung_gio = khung_gio.sort_values(by='Giờ').reset_index(drop=True)

print("Số khách quay lại theo từng giờ (định dạng khung giờ):")
print(khung_gio)

# Danh sách các khung giờ bạn cần lấy, tách riêng từng giờ
khung_gio_can_lay = [
    '10:00-10:59',
    '18:00-18:59',
    '21:00-21:59'
]

# Lọc từ bảng khung_gio
result = khung_gio[khung_gio['Giờ'].isin(khung_gio_can_lay)].reset_index(drop=True)

print(result)

Số khách quay lại theo từng giờ (định dạng khung giờ):
            Giờ  Số khách quay lại
0   10:00-10:59                 25
1   11:00-11:59                  9
2   12:00-12:59                 21
3   13:00-13:59                  9
4   14:00-14:59                 11
5   15:00-15:59                 19
6   16:00-16:59                 21
7   17:00-17:59                 19
8   18:00-18:59                 26
9   19:00-19:59                 18
10  20:00-20:59                 20
11  21:00-21:59                 26
12  22:00-22:59                 12
13  23:00-23:59                 21
14    8:00-8:59                 15
15    9:00-9:59                 16
           Giờ  Số khách quay lại
0  10:00-10:59                 25
1  18:00-18:59                 26
2  21:00-21:59                 26


#IV. PHÂN TÍCH MÙA VỤ

##4.1. Dashboard phân tích mùa vụ: Sự tăng trưởng doanh thu theo từng tháng

Biểu đồ cho thấy doanh thu có sự biến động rõ ràng theo thời gian, từ đầu năm đến tháng 7 tăng trưởng doanh thu giai đoạn này khá thấp thậm chí có những tháng tăng trưởng âm với cùng kỳ tháng trước, đặc biệt là vào thời điểm cuối năm từ tháng 8 có sự bùng nổ doanh thu tăng trưởng lên đến 91.2% so với tháng trước đó và đạt đỉnh trong năm vào tháng 12 với doanh thu lên đến 750 triệu VND. để hiểu hơn tại sao thì đi vào tìm hiểu về từng nhóm hàng và từng mặt hàng theo tháng


In [19]:
# Chuyển 'Thời gian tạo đơn' sang datetime nếu chưa
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

# Tạo cột 'tháng' theo định dạng "Tháng 01", "Tháng 02"...
df['tháng'] = 'Tháng ' + df['Thời gian tạo đơn'].dt.month.apply(lambda x: f"{x:02d}")

# Nhóm theo 'tháng' và tính tổng doanh thu từng tháng
df_monthly = df.groupby('tháng')['Thành tiền'].sum().reset_index()

# Đổi tên cột tổng doanh thu thành 'Doanh thu'
df_monthly = df_monthly.rename(columns={'Thành tiền': 'Doanh thu'})

# Tính MoM% theo công thức:
# (Doanh thu tháng hiện tại - Doanh thu tháng trước) / |Doanh thu tháng trước| * 100
df_monthly['MoM%'] = (df_monthly['Doanh thu'] - df_monthly['Doanh thu'].shift(1)) / df_monthly['Doanh thu'].shift(1).abs() * 100

# Hiển thị kết quả
df_monthly

tháng  Doanh thu       MoM%
0   Tháng 01  285858000        NaN
1   Tháng 02  294731000   3.103989
2   Tháng 03  257784000 -12.535838
3   Tháng 04  196365000 -23.825761
4   Tháng 05  214801000   9.388639
5   Tháng 06  262975000  22.427270
6   Tháng 07  271803000   3.356973
7   Tháng 08  519576000  91.159038
8   Tháng 09  528032000   1.627481
9   Tháng 10  567778000   7.527195
10  Tháng 11  629286000  10.833107
11  Tháng 12  750085000  19.196200

##4.2. Dashboard phân tích mùa vụ: Biến động doanh thu của từng mặt hàng theo từng tháng

Như vậy, nhìn chung các mặt hàng thuộc các nhóm hàng đều có xu hướng doanh thu ổn định hoặc tăng giảm nhẹ trong nửa đầu năm, sau đó bùng nổ mạnh từ tháng 8 trở đi, đặc biệt là vào các tháng cuối năm.
Đây có thể do yếu tố mùa vụ theo thời tiết: Tháng 8 trở đi thường bắt đầu mùa thu - đông, khí hậu mát mẻ hơn, đây là thời điểm nhu cầu tiêu thụ các sản phẩm trà nóng và thực phẩm chế biến từ bột (như bột ăn, thức uống nóng) tăng cao hơn.
Hoặc do các yếu tố mùa vụ theo các dịp lễ lớn trong năm: cuối năm doanh thu của các sản phẩm trà, bột bùng nổ có thể do cuối năm là mùa cao điểm mua sắm, quà tặng, chuẩn bị Tết Nguyên đán.Trà và các sản phẩm liên quan thường được chọn làm quà biếu, quà tặng trong dịp lễ hội, dẫn đến doanh thu tăng mạnh.

In [22]:
# Chuyển 'Thời gian tạo đơn' sang datetime nếu chưa
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

# Tạo cột 'tháng' theo định dạng "Tháng 01", "Tháng 02"...
df['tháng'] = 'Tháng ' + df['Thời gian tạo đơn'].dt.month.apply(lambda x: f"{x:02d}")

# Nhóm dữ liệu theo 'tháng', 'Mặt hàng', 'Nhóm hàng', tính tổng 'Thành tiền'
df_summary = df.groupby(['tháng', 'Tên mặt hàng', 'Tên nhóm hàng'], as_index=False)['Thành tiền'].sum()

# Đổi tên cột thành 'Doanh thu'
df_summary = df_summary.rename(columns={'Thành tiền': 'Doanh thu'})

# Hiển thị kết quả
df_summary

tháng                 Tên mặt hàng    Tên nhóm hàng  Doanh thu
0    Tháng 01                  Bột cần tây              Bột   36000000
1    Tháng 01                      Cam lát  Trà củ, quả sấy   22365000
2    Tháng 01    Set 10 gói trà cam sả quế          Set trà    4189000
3    Tháng 01    Set 10 gói trà dưỡng nhan          Set trà    9165000
4    Tháng 01  Set 10 gói trà gạo lứt 8 vị          Set trà    2695000
..        ...                          ...              ...        ...
223  Tháng 12            Trà hoa cúc trắng          Trà hoa   60320000
224  Tháng 12             Trà hoa đậu biếc          Trà hoa   41072000
225  Tháng 12        Trà nhuỵ hoa nghệ tây          Trà hoa   86580000
226  Tháng 12     Trà nụ hoa hồng Tây Tạng          Trà hoa   21560000
227  Tháng 12        Trà nụ hoa nhài trắng          Trà hoa   41405000

[228 rows x 4 columns]

##4.3. Dashboard phân tích mùa vụ: Doanh thu trung bình theo ngày trong tuần và tháng.

Và dựa vào chart heatmap này ta thấy doanh thu thường tăng cao hơn vào thời điểm cuối tuần trong trong năm (từ thứ 6 đến chủ nhật) và càng tăng cao về giai đoạn cuối năm, thứ 7 chủ nhật gần như luôn là hai ngày có doanh thu cao nhất trong tuần ở tất cả các tháng như Tháng 8 doanh thu Thứ Bảy khoảng 18.9 triệu, Chủ Nhật khoảng 18.3 triệu, cao hơn rất nhiều so với các ngày khác trong tuần.
Điều này cho thấy khách hàng ưu tiên mua vào cuối tuần, có thể do thời gian rảnh rỗi, các hoạt động giải trí, mua sắm gia đình.
Tập trung đẩy mạnh hoạt động bán hàng vào cuối tuần đặc biệt là vào giai đoạn cuối năm để tận dụng cao điểm mùa vụ. Tạo ra các chương trình thu hút khách hàng tận dụng tâm lý mua sắm của khách hàng vào các ngày này. Xem xét mở rộng kênh bán hàng đa phương tiện để dễ tiếp cận khách hàng hơn trong các ngày giữa tuần, đẩy mạnh Marketing về những tác động tích cực của các loại trà, bột đối với sức khỏe con người, thúc đẩy doanh thu trong các tháng đầu năm bằng chương trình kích cầu để giảm thiểu giai đoạn giảm sút doanh thu vào những năm kế tiếp

In [23]:
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])

df['tháng'] = 'Tháng ' + df['Thời gian tạo đơn'].dt.month.apply(lambda x: f"{x:02d}")

weekday_map = {
    'Sunday': 'Chủ Nhật',
    'Monday': 'Thứ Hai',
    'Tuesday': 'Thứ Ba',
    'Wednesday': 'Thứ Tư',
    'Thursday': 'Thứ Năm',
    'Friday': 'Thứ Sáu',
    'Saturday': 'Thứ Bảy'
}
df['weekday_en'] = df['Thời gian tạo đơn'].dt.day_name()
df['ngày trong tuần'] = df['weekday_en'].map(weekday_map)
df.drop(columns=['weekday_en'], inplace=True)

# Tạo pivot table: index là 'ngày trong tuần', columns là 'tháng', giá trị là tổng 'Thành tiền'
pivot_df = pd.pivot_table(
    df,
    values='Thành tiền',
    index='ngày trong tuần',
    columns='tháng',
    aggfunc='sum',
    fill_value=0
)

# Sắp xếp thứ tự ngày trong tuần cho đúng (từ Thứ Hai đến Chủ Nhật)
thu_tu_ngay = ['Thứ Hai', 'Thứ Ba', 'Thứ Tư', 'Thứ Năm', 'Thứ Sáu', 'Thứ Bảy', 'Chủ Nhật']
pivot_df = pivot_df.reindex(thu_tu_ngay)

# Sắp xếp cột tháng theo thứ tự từ Tháng 01 đến Tháng 12
pivot_df = pivot_df.reindex(sorted(pivot_df.columns), axis=1)

pivot_df

tháng            Tháng 01  Tháng 02  Tháng 03  Tháng 04  Tháng 05  Tháng 06  \
ngày trong tuần                                                               
Thứ Hai          41115000  39035000  30131000  25823000  29100000  30631000   
Thứ Ba           31979000  41946000  34181000  23855000  31035000  33529000   
Thứ Tư           33831000  36953000  40228000  24500000  26119000  43894000   
Thứ Năm          38872000  45220000  43597000  25325000  25655000  42001000   
Thứ Sáu          41525000  42512000  36716000  33807000  30247000  39760000   
Thứ Bảy          53106000  45257000  35152000  35427000  32814000  38708000   
Chủ Nhật         45430000  43808000  37779000  27628000  39831000  34452000   

tháng            Tháng 07  Tháng 08  Tháng 09  Tháng 10   Tháng 11   Tháng 12  
ngày trong tuần                                                                
Thứ Hai          31610000  79122000  68739000  79518000   77943000   85838000  
Thứ Ba           28670000  79751000  65939000  66824000   93383000   87280000  
Thứ Tư           34334000  77734000  68122000  72098000  104344000   83802000  
Thứ Năm          35813000  66081000  80697000  77161000   87247000  119067000  
Thứ Sáu          43056000  67880000  87876000  77927000   86416000  136803000  
Thứ Bảy          50495000  75774000  77351000  98308000   91196000  134570000  
Chủ Nhật         47825000  73234000  79308000  95942000   88757000  102725000

##4.4. Dashboard phân tích mùa vụ: Doanh thu trung bình theo ngày trong tuần và tháng.

Biểu đồ cho thấy nhóm khách hàng A1 - bao gồm nhân viên văn phòng, chủ doanh nghiệp từ 36–45 tuổi - có hành vi chi tiêu rõ rệt theo mùa: doanh thu thấp nhất từ tháng 4 đến tháng 7, nhưng bùng nổ mạnh mẽ từ tháng 8 đến cuối năm, tăng gấp nhiều lần. Điều này gợi ý rằng cần có chiến dịch kích cầu sớm cho nhóm này trong mùa thấp điểm – chẳng hạn như ưu đãi độc quyền, combo sức khỏe, hoặc khuyến mãi thành viên. Trong khi đó, các nhóm A2, A3 có xu hướng tăng đều từ giữa năm trở đi, phù hợp để triển khai chương trình “mua sớm – giá tốt” từ tháng 6–7. Với nhóm khách hàng thấp hơn, nên tập trung vào sản phẩm thử nghiệm, giá rẻ để tăng mức tiếp cận thay vì đẩy mạnh ngân sách quảng bá

In [24]:
df['Thời gian tạo đơn'] = pd.to_datetime(df['Thời gian tạo đơn'])
df['tháng'] = 'Tháng ' + df['Thời gian tạo đơn'].dt.month.apply(lambda x: f"{x:02d}")

# Tạo pivot table: index là 'tháng', columns là 'Mã PKKH', giá trị là tổng 'Thành tiền'
pivot_ma_kh = pd.pivot_table(
    df,
    values='Thành tiền',
    index='tháng',
    columns='Mã PKKH',
    aggfunc='sum',
    fill_value=0
)

# Sắp xếp index tháng theo thứ tự từ Tháng 01 đến Tháng 12
pivot_ma_kh = pivot_ma_kh.reindex(sorted(pivot_ma_kh.index), axis=0)

pivot_ma_kh

Mã PKKH          A1         A2        A3         B1        B2        B3  \
tháng                                                                     
Tháng 01   67922000   41698000  22055000   38913000  36466000  16789000   
Tháng 02   66971000   55475000  14842000   52529000  42430000  13021000   
Tháng 03   63228000   47147000  11984000   55587000  30088000  13612000   
Tháng 04   29393000   30475000  16708000   42523000  34860000  16088000   
Tháng 05   34691000   37209000  16288000   42860000  38086000  18626000   
Tháng 06   36347000   46706000  20055000   51177000  51364000  22679000   
Tháng 07   38745000   46329000  22862000   55647000  51243000  25190000   
Tháng 08  154106000   92996000  39462000   92305000  48841000  26543000   
Tháng 09  142074000   88619000  29515000  108975000  71150000  26999000   
Tháng 10  148853000   98588000  38434000  106899000  76791000  29922000   
Tháng 11  167056000  111517000  38140000  119542000  93543000  32647000   
Tháng 12  192695000  132239000  43400000  138741000  99917000  46375000   

Mã PKKH         C1        C2        C3  
tháng                                   
Tháng 01  35689000  16033000  10293000  
Tháng 02  28240000  13527000   7696000  
Tháng 03  17434000  10911000   7793000  
Tháng 04   8229000  10590000   7499000  
Tháng 05   8587000  11198000   7256000  
Tháng 06  11012000  14154000   9481000  
Tháng 07  10263000  14184000   7340000  
Tháng 08  22697000  25259000  17367000  
Tháng 09  23578000  21381000  15741000  
Tháng 10  26916000  20553000  20822000  
Tháng 11  23784000  23627000  19430000  
Tháng 12  42459000  32296000  21963000